File: python/socket_utils.py
"""
Socket path utilities for shared daemon management.
Converts model IDs to safe socket filenames for Unix domain sockets.
"""

import re
from pathlib import Path

MLX_COMMUNITY_PREFIX = "mlx-community/"
DAEMON_DIR_NAME = ".mlx-hub"
DAEMONS_SUBDIR = "daemons"


def model_id_to_socket_name(model_id: str) -> str:
    """
    Convert a model ID to a safe socket filename.

    Transformation rules:
    1. Strip 'mlx-community/' prefix if present
    2. Convert to lowercase
    3. Replace non-alphanumeric characters with hyphens
    4. Collapse multiple consecutive hyphens
    5. Remove leading and trailing hyphens

    Example:
        >>> model_id_to_socket_name('mlx-community/DeepSeek-Coder-V2-Lite-Instruct-4bit')
        'deepseek-coder-v2-lite-instruct-4bit'
    """
    name = model_id

    # Strip mlx-community/ prefix if present
    if name.startswith(MLX_COMMUNITY_PREFIX):
        name = name[len(MLX_COMMUNITY_PREFIX):]

    # Lowercase
    name = name.lower()

    # Replace non-alphanumeric with hyphens
    name = re.sub(r"[^a-z0-9]", "-", name)

    # Collapse multiple hyphens
    name = re.sub(r"-+", "-", name)

    # Remove leading/trailing hyphens
    name = name.strip("-")

    return name


def get_daemon_dir() -> Path:
    """
    Get the directory where daemon sockets and pid files are stored.

    Returns:
        Path to ~/.mlx-hub/daemons
    """
    return Path.home() / DAEMON_DIR_NAME / DAEMONS_SUBDIR


def get_socket_path(model_id: str) -> Path:
    """
    Get the Unix domain socket path for a model.

    Args:
        model_id: The model ID (e.g., 'mlx-community/Llama-3.2-1B-Instruct-4bit')

    Returns:
        Path to the socket file (e.g., ~/.mlx-hub/daemons/llama-3-2-1b-instruct-4bit.sock)
    """
    socket_name = model_id_to_socket_name(model_id)
    return get_daemon_dir() / f"{socket_name}.sock"


def get_pid_path(model_id: str) -> Path:
    """
    Get the PID file path for a model's daemon.

    Args:
        model_id: The model ID (e.g., 'mlx-community/Llama-3.2-1B-Instruct-4bit')

    Returns:
        Path to the PID file (e.g., ~/.mlx-hub/daemons/llama-3-2-1b-instruct-4bit.pid)
    """
    socket_name = model_id_to_socket_name(model_id)
    return get_daemon_dir() / f"{socket_name}.pid"


if __name__ == "__main__":
    # Quick test
    test_cases = [
        "mlx-community/DeepSeek-Coder-V2-Lite-Instruct-4bit",
        "mlx-community/Llama-3.2-1B-Instruct-4bit",
        "some-org/Some-Model-4bit",
    ]

    for model_id in test_cases:
        print(f"Model ID: {model_id}")
        print(f"  Socket name: {model_id_to_socket_name(model_id)}")
        print(f"  Socket path: {get_socket_path(model_id)}")
        print(f"  PID path: {get_pid_path(model_id)}")
        print()

File: src/socket-utils.ts
/**
 * Socket path utilities for shared daemon management.
 * Converts model IDs to safe socket filenames for Unix domain sockets.
 */

import { homedir } from 'os';
import { join } from 'path';
import { existsSync, readdirSync, readFileSync } from 'fs';

const MLX_COMMUNITY_PREFIX = 'mlx-community/';
const DAEMON_DIR_NAME = '.mlx-hub';
const DAEMONS_SUBDIR = 'daemons';

/**
 * Convert a model ID to a safe socket filename.
 *
 * Transformation rules:
 * 1. Strip 'mlx-community/' prefix if present
 * 2. Convert to lowercase
 * 3. Replace non-alphanumeric characters with hyphens
 * 4. Collapse multiple consecutive hyphens
 * 5. Remove leading and trailing hyphens
 *
 * @example
 * modelIdToSocketName('mlx-community/DeepSeek-Coder-V2-Lite-Instruct-4bit')
 * // => 'deepseek-coder-v2-lite-instruct-4bit'
 */
export function modelIdToSocketName(modelId: string): string {
  // Strip mlx-community/ prefix if present
  let name = modelId;
  if (name.startsWith(MLX_COMMUNITY_PREFIX)) {
    name = name.slice(MLX_COMMUNITY_PREFIX.length);
  }

  // Lowercase
  name = name.toLowerCase();

  // Replace non-alphanumeric with hyphens
  name = name.replace(/[^a-z0-9]/g, '-');

  // Collapse multiple hyphens
  name = name.replace(/-+/g, '-');

  // Remove leading/trailing hyphens
  name = name.replace(/^-+|-+$/g, '');

  return name;
}

/**
 * Get the directory where daemon sockets and pid files are stored.
 *
 * @returns Path to ~/.mlx-hub/daemons
 */
export function getDaemonDir(): string {
  return join(homedir(), DAEMON_DIR_NAME, DAEMONS_SUBDIR);
}

/**
 * Get the Unix domain socket path for a model.
 *
 * @param modelId - The model ID (e.g., 'mlx-community/Llama-3.2-1B-Instruct-4bit')
 * @returns Path to the socket file (e.g., ~/.mlx-hub/daemons/llama-3-2-1b-instruct-4bit.sock)
 */
export function getSocketPath(modelId: string): string {
  const socketName = modelIdToSocketName(modelId);
  return join(getDaemonDir(), `${socketName}.sock`);
}

/**
 * Get the PID file path for a model's daemon.
 *
 * @param modelId - The model ID (e.g., 'mlx-community/Llama-3.2-1B-Instruct-4bit')
 * @returns Path to the PID file (e.g., ~/.mlx-hub/daemons/llama-3-2-1b-instruct-4bit.pid)
 */
export function getPidPath(modelId: string): string {
  const socketName = modelIdToSocketName(modelId);
  return join(getDaemonDir(), `${socketName}.pid`);
}

/**
 * Information about a running daemon.
 */
export interface RunningDaemon {
  socketName: string;
  socketPath: string;
  pid: number;
  isAlive: boolean;
}

/**
 * List all running daemons by scanning the daemon directory.
 * Checks both socket files and PID files, verifying processes are alive.
 *
 * @returns Array of running daemon info
 */
export function listRunningDaemons(): RunningDaemon[] {
  const daemonDir = getDaemonDir();

  if (!existsSync(daemonDir)) {
    return [];
  }

  const files = readdirSync(daemonDir);
  const socketFiles = files.filter((f) => f.endsWith('.sock'));
  const daemons: RunningDaemon[] = [];

  for (const sockFile of socketFiles) {
    const socketName = sockFile.replace('.sock', '');
    const socketPath = join(daemonDir, sockFile);
    const pidPath = join(daemonDir, `${socketName}.pid`);

    let pid = 0;
    let isAlive = false;

    if (existsSync(pidPath)) {
      try {
        pid = parseInt(readFileSync(pidPath, 'utf-8').trim(), 10);
        if (!isNaN(pid)) {
          // Send signal 0 to check if process exists
          process.kill(pid, 0);
          isAlive = true;
        }
      } catch {
        // Process doesn't exist or permission denied
        isAlive = false;
      }
    }

    daemons.push({ socketName, socketPath, pid, isAlive });
  }

  return daemons;
}
